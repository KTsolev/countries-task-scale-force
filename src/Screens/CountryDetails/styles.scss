.details {
    @include column;
    padding: 12px;
    position: relative;

    &-item {
        @include column;
        justify-content: space-around;
        align-items: stretch;

        .svg {
            position: absolute;
            left: 10px;
            top: 10px;
            padding: 2px 4px;
            width: 30px;
            height: 30px;
        }

        &--row {
            @include row;
            padding: 12px;
            background-color: #d0eaeb;
        }

        &-label,
        &-header {
            @include column;
            border: 1px ridge #b9b9b9;
            align-items: center;
        }

        &-label {
            font-size: 18px;
            color: #4a4b4a;
        }

        &-header {
            font-size: 18px;
            color: #353535;
            font-weight: bold;
        }
    }
}

.column {
     flex-direction: column !important;
    justify-content: space-around !important;
    align-items: stretch !important;
    background-color: transparent !important;
}

