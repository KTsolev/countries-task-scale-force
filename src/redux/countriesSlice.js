import { createSlice } from '@reduxjs/toolkit';

const initialState  = {
    loading: true,
    error: null,
    selectedCountry: null,
    result: []
  }
  
  export const counterSlice = createSlice({
    name: 'countries',
    // `createSlice` will infer the state type from the `initialState` argument
    initialState,
    reducers: {
      fetchCountries: (state, action) => {
        return {
          ...state,
          error: null,
          loading: true
        };
      },
      fetchCountriesSuccess: (state, action) => {
        return {
            ...state,
            loading: false,
            result: action.payload
        };
      },
      fetchCountriesFailure: (state, action) => {
        return {
            ...state,
            loading: false,
            error: action.payload
        };
      },
      cancelUserAction: (state, action) => {
        return {
          ...state,
          selectedCountry: null
        };
      },
      beginUserAction: (state, action) => {
        return {
          ...state,
        };
      },
      selectCountry: (state, action) => {
        return {
          ...state,
          selectedCountry: action.payload,
        };
      },
      fetchCountryByName: (state, action) => {
        return {
          ...state,
          loading: true
        };
      },
    },
  });

export const { 
  fetchCountries, 
  fetchCountriesSuccess,
  fetchCountriesFailure,
  selectCountry,
  fetchCountryByName,
  cancelUserAction,
  beginUserAction
 } = counterSlice.actions;

export const countriesReducer = counterSlice.reducer;