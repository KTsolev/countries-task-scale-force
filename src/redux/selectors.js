export const selectCountries = (state) => state.countries.result;
export const selectedCountry = (state) => state.countries.selectedCountry;
export const isLoading = (state) => state.countries.loading;
export const error = (state) => state.countries.error;

