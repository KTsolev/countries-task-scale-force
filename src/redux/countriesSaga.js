import { call, put, takeLatest, select, take, race, spawn, delay } from 'redux-saga/effects';
import { getCountries } from '../network/endpoints';
import { 
    fetchCountries,
    fetchCountriesSuccess,
    fetchCountriesFailure,
    cancelUserAction,
    beginUserAction,
    selectCountry
} from './countriesSlice';
import { selectCountries } from './selectors';

export function* watchCountriesSaga () {
    yield takeLatest(fetchCountries.type, fetchCountriesInit);
}

function* fetchCountriesInit(action) {
    try {
      const { page, itemsPerPage, reload } = action.payload;
      const countries = yield select(selectCountries);
      let result;
      
      if (countries.result && !reload ) {
         result = { ...countries };
         yield put(fetchCountriesSuccess({ ...result }));
      } else {
         result = yield call(getCountries,  page, itemsPerPage);
         yield put(fetchCountriesSuccess({...result.data[0], items: result.data[1] }));
      }

   } catch (error) {
      yield put(fetchCountriesFailure(error));
   }
}

export function* watchUserAction() {
    while (true) {
       let action = yield take(beginUserAction.type);
       yield spawn(bgTask, action);
    }
 }
 
 function* bgTask (task) {
    const selected = task.payload;
    const { undo } = yield race({
       undo: take((action) => {
          return action.type === cancelUserAction.type;
       }),
       proceed: delay(3000)
    });
    
    if (undo) {
       yield put(cancelUserAction(null));
    } else {
       yield put(selectCountry(selected));
    }
 }