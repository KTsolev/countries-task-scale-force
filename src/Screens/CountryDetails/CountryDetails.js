import React, { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { useHistory } from "react-router-dom";
import { CountryTile } from '../../shared_ui_components/CountryTile/CountryTile';
import { selectedCountry } from '../../redux/selectors';
import {ReactComponent as BackArrow} from '../../assets/back_arrow_left.svg';
import { cancelUserAction } from '../../redux/countriesSlice';

export const CountryDetails  = (props) => {
    const country = useSelector(selectedCountry);
    const history = useHistory();
    const dispatch = useDispatch();

    useEffect(() => {
        return () => {
           dispatch(cancelUserAction(null));
        }
    }, [])

    const goBack = () => {
        history.push('/');
    }   
    
    if (!country) {
        goBack();
        return null;
    }

    return (
        <section className='details'>
            <h2>Country Details</h2>
            <div className='details-item' onClick={goBack}>
                <BackArrow className='svg' />
            </div>
            <div className='details-item details-item--row'>
                <div className='details-item'>
                    <span className='details-item-header'>Iso Code</span>
                    <span className='details-item-header'>Name</span>    
                    <span className='details-item-header'>CapitalCity</span>    
                    <span className='details-item-header'>Longitude</span>  
                    <span className='details-item-header'>Latitude</span>
                </div>
                <CountryTile country={country} dissableActions cssClassName='column'/>
            </div>
            <div className='details-item details-item--row'>
                <div className='details-item'>
                    <span className='details-item-header'>Region</span>
                </div>
                <div className='details-item'>
                    <span className='details-item-label'>{country.region.value || 'N/A'}</span>
                    <span className='details-item-label'>{country.region.id || 'N/A'}</span>
                    <span className='details-item-label'>{country.region.iso2code || 'N/A'}</span>
                </div>
            </div>
            <div className='details-item details-item--row'>
                <div className='details-item'>
                    <span className='details-item-header'>Adminregion</span>
                </div>
                <div className='details-item'>
                    <span className='details-item-label'>{country.adminregion.id || 'N/A'}</span>
                    <span className='details-item-label'>{country.adminregion.iso2code || 'N/A'}</span>
                    <span className='details-item-label'>{country.adminregion.value || 'N/A'}</span>
                </div>
            </div>
            <div className='details-item details-item--row'>
                <div className='details-item'>
                    <span className='details-item-header'>IncomeLevel</span>
                </div>
                <div className='details-item'>
                    <span className='details-item-label'>{country.incomeLevel.id || 'N/A'}</span>
                    <span className='details-item-label'>{country.incomeLevel.iso2code || 'N/A'}</span>
                    <span className='details-item-label'>{country.incomeLevel.value || 'N/A'}</span>
                </div>
            </div>
            <div className='details-item details-item--row'>
                <div className='details-item'>
                    <span className='details-item-header'>LendingType</span>
                </div>
                <div className='details-item'>
                    <span className='details-item-label'>{country.lendingType.id || 'N/A'}</span>
                    <span className='details-item-label'>{country.lendingType.iso2Code || 'N/A'}</span>
                    <span className='details-item-label'>{country.lendingType.value || 'N/A'}</span>
                </div>
            </div>
        </section>
    );
}