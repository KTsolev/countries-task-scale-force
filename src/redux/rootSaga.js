import { all } from 'redux-saga/effects';
import { watchCountriesSaga, watchUserAction } from '../Screens/Countries/countriesSaga';

export function* rootSaga() {
    yield all([
        watchCountriesSaga(),
        watchUserAction()
    ]);
};