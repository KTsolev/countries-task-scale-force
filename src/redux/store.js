import logger from 'redux-logger';
import { configureStore } from '@reduxjs/toolkit';
import { countriesReducer } from './countriesSlice';
import { rootSaga } from './rootSaga';
import createSagaMiddleware from 'redux-saga';

const sagaMiddleware = createSagaMiddleware();

export const store = configureStore({
  reducer: {
    countries: countriesReducer,
  },
  middleware: [logger, sagaMiddleware]
});

sagaMiddleware.run(rootSaga);