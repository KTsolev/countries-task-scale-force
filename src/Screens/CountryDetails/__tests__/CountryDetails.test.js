import React from 'react';
import { CountryDetails } from '../CountryDetails';
import renderer, { act } from 'react-test-renderer'; 
import { selectedCountry } from '../../../redux/selectors';
// eslint-disable-next-line jest/no-mocks-import
import jsonCountries from '../../../__mocks__/mockedParsedJSON';

const mockedHistory = jest.fn();
const mockDispatchFn = jest.fn();
const mockedCountries = jsonCountries;

jest.mock('react-router-dom', () => ({
  useHistory: () => ({
    push: jest.fn(),
    replace: mockedHistory,
  }),
}));

jest.mock('react-redux', () => ({
  useSelector: selector => selector(),
  useDispatch: () => mockDispatchFn
}));

jest.mock('../../../redux/selectors.js', () => ({
    selectedCountry: jest.fn()
}));

describe('CountryDetails tests', () => {
      
    afterEach(() => {
      jest.resetAllMocks();
    });
    
    it('should not render country details when no selected country and should navigate to main screen', () => {
       selectedCountry.mockReturnValueOnce(null);
       const tree = renderer.create(<CountryDetails />);
       expect(mockedHistory).toHaveBeenCalled(); 
       expect(tree.toJSON()).toMatchSnapshot();
    });

    it('should render country details', () => {
      selectedCountry.mockReturnValueOnce(mockedCountries.items[4]);

      const tree = renderer.create(<CountryDetails />);
      expect(tree.toJSON()).toMatchSnapshot();
    });

   it('should redirect to main screen when back button is clicked', () => {
      selectedCountry.mockReturnValueOnce(mockedCountries.items[4]);
      const tree = renderer.create(<CountryDetails />).root;
      const button = tree.findAllByProps({className: 'details-item'})[0];

      button.props.onClick();
      expect(mockedHistory).toHaveBeenCalled(); 
   });

   it('should call cleanup function when unmounted', () => {
      selectedCountry.mockReturnValueOnce(mockedCountries.items[4]);
      jest.spyOn(React, "useEffect").mockImplementationOnce(cb => cb()());
      
      const action = {"payload": null, "type": "countries/cancelUserAction"};
      const tree = renderer.create(<CountryDetails />);

      act(tree.unmount);
    
      expect(mockDispatchFn).toHaveBeenCalledWith(action); 
  });
});
