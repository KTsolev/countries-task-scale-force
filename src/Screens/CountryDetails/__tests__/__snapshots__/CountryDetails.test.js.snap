// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CountryDetails tests should not render country details when no selected country and should navigate to main screen 1`] = `null`;

exports[`CountryDetails tests should render country details 1`] = `
<section
  className="details"
>
  <h2>
    Country Details
  </h2>
  <div
    className="details-item"
    onClick={[Function]}
  >
    <test
      className="svg"
    />
  </div>
  <div
    className="details-item details-item--row"
  >
    <div
      className="details-item"
    >
      <span
        className="details-item-header"
      >
        Iso Code
      </span>
      <span
        className="details-item-header"
      >
        Name
      </span>
      <span
        className="details-item-header"
      >
        CapitalCity
      </span>
      <span
        className="details-item-header"
      >
        Longitude
      </span>
      <span
        className="details-item-header"
      >
        Latitude
      </span>
    </div>
    <div
      className="column country-tile"
      onMouseDownCapture={[Function]}
      onMouseUpCapture={[Function]}
    >
      <div
        className="country-tile-inner"
      />
      <span
        className="country-tile-item"
      >
        ZI
      </span>
      <span
        className="country-tile-item"
      >
        Africa Western and Central
      </span>
      <span
        className="country-tile-item"
      >
        N/A
      </span>
      <span
        className="country-tile-item"
      >
        N/A
      </span>
      <span
        className="country-tile-item"
      >
        N/A
      </span>
    </div>
  </div>
  <div
    className="details-item details-item--row"
  >
    <div
      className="details-item"
    >
      <span
        className="details-item-header"
      >
        Region
      </span>
    </div>
    <div
      className="details-item"
    >
      <span
        className="details-item-label"
      >
        Aggregates
      </span>
      <span
        className="details-item-label"
      >
        NA
      </span>
      <span
        className="details-item-label"
      >
        NA
      </span>
    </div>
  </div>
  <div
    className="details-item details-item--row"
  >
    <div
      className="details-item"
    >
      <span
        className="details-item-header"
      >
        Adminregion
      </span>
    </div>
    <div
      className="details-item"
    >
      <span
        className="details-item-label"
      >
        N/A
      </span>
      <span
        className="details-item-label"
      >
        N/A
      </span>
      <span
        className="details-item-label"
      >
        N/A
      </span>
    </div>
  </div>
  <div
    className="details-item details-item--row"
  >
    <div
      className="details-item"
    >
      <span
        className="details-item-header"
      >
        IncomeLevel
      </span>
    </div>
    <div
      className="details-item"
    >
      <span
        className="details-item-label"
      >
        NA
      </span>
      <span
        className="details-item-label"
      >
        NA
      </span>
      <span
        className="details-item-label"
      >
        Aggregates
      </span>
    </div>
  </div>
  <div
    className="details-item details-item--row"
  >
    <div
      className="details-item"
    >
      <span
        className="details-item-header"
      >
        LendingType
      </span>
    </div>
    <div
      className="details-item"
    >
      <span
        className="details-item-label"
      >
        N/A
      </span>
      <span
        className="details-item-label"
      >
        N/A
      </span>
      <span
        className="details-item-label"
      >
        Aggregates
      </span>
    </div>
  </div>
</section>
`;
